<a href="https://www.kaggle.com/code/ahmedfakhar123/ml-43-the-naive-bayes-classifier?scriptVersionId=346534641" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

<div align="center">
<div style="background: linear-gradient(135deg, #1a0533 0%, #2d1060 40%, #4a1a8a 70%, #6b2fa0 100%); border-radius: 20px; padding: 60px 40px 50px; margin-bottom: 10px; position: relative; overflow: hidden;">
<div style="position: absolute; top: 0; left: 0; right: 0; bottom: 0; background: radial-gradient(ellipse at 70% 30%, rgba(168,85,247,0.25) 0%, transparent 65%); pointer-events: none;"></div>
<div style="font-size: 13px; font-weight: 600; letter-spacing: 3px; color: #c084fc; text-transform: uppercase; margin-bottom: 20px;">Episode 43 · Classification</div>
<h1 style="font-size: 52px; font-weight: 800; color: #ffffff; margin: 0 0 16px; line-height: 1.1; text-shadow: 0 2px 20px rgba(168,85,247,0.4);">🔮 Naive Bayes</h1>
 
<p style="font-size: 20px; color: #e9d5ff; margin: 0 0 32px; font-weight: 300; max-width: 560px; margin-left: auto; margin-right: auto; line-height: 1.6;">Five flavors. One powerful probabilistic idea. Learn which variant to reach for — and why.</p>
<div style="display: flex; justify-content: center; gap: 12px; flex-wrap: wrap;">
  <span style="background: rgba(168,85,247,0.25); border: 1px solid rgba(168,85,247,0.5); color: #e9d5ff; padding: 7px 18px; border-radius: 50px; font-size: 13px;">🌸 Iris</span>
  <span style="background: rgba(168,85,247,0.25); border: 1px solid rgba(168,85,247,0.5); color: #e9d5ff; padding: 7px 18px; border-radius: 50px; font-size: 13px;">📱 SMS Spam</span>
  <span style="background: rgba(168,85,247,0.25); border: 1px solid rgba(168,85,247,0.5); color: #e9d5ff; padding: 7px 18px; border-radius: 50px; font-size: 13px;">🚢 Titanic</span>
  <span style="background: rgba(168,85,247,0.25); border: 1px solid rgba(168,85,247,0.5); color: #e9d5ff; padding: 7px 18px; border-radius: 50px; font-size: 13px;">sklearn</span>
</div>
</div>
</div>


---
 
## 📋 Table of Contents
 
| # | Section |
|---|---------|
| 1 | [What is Naive Bayes?](#naive-bayes) |
| 2 | [Bayes' Theorem](#bayes) |
| 3 | [Why "Naive"?](#naive) |
| 4 | [GaussianNB — Iris Dataset](#gaussian) |
| 5 | [MultinomialNB — SMS Spam](#multinomial) |
| 6 | [BernoulliNB — SMS Spam](#bernoulli) |
| 7 | [CategoricalNB — Titanic](#categorical) |
| 8 | [ComplementNB — SMS Spam](#complement) |
| 9 | [Comparison & Cheat Sheet](#cheatsheet) |
 
---

<a id="naive-bayes"></a>
 
<div style="background: linear-gradient(135deg, #1a0533, #2d1060); border-left: 5px solid #a855f7; border-radius: 12px; padding: 28px 32px; margin: 30px 0;">
<h2 style="color: #e9d5ff; margin: 0 0 14px; font-size: 24px;">🔮 What is Naive Bayes?</h2>
<p style="color: #d8b4fe; margin: 0; line-height: 1.8; font-size: 15px;">
Naive Bayes is a family of <strong style="color: #f3e8ff;">probabilistic classifiers</strong> built on a single idea: use known probabilities to predict unknown outcomes. Instead of learning a complex decision boundary like a neural network or SVM, it asks a simpler question — <em>"Given what I've seen in the training data, how likely is each class?"</em>
<br><br>
It's fast, interpretable, and often surprisingly effective — especially on text and high-dimensional data. Despite its simplicity, Naive Bayes is a staple in spam filters, sentiment classifiers, and medical diagnosis tools.
</p>
</div>

---
 
<a id="bayes"></a>
 
## 📐 Bayes' Theorem
 
Everything in this notebook flows from one equation discovered by Thomas Bayes in the 18th century:
 
<div style="background: #0f0520; border: 1px solid #6b21a8; border-radius: 14px; padding: 30px 24px; margin: 24px 0; text-align: center;">
<div style="font-size: 28px; font-weight: 700; color: #e9d5ff; font-family: 'Georgia', serif; letter-spacing: 1px; margin-bottom: 20px;">
P(y | X) = <span style="color: #c084fc;">P(X | y) · P(y)</span> / <span style="color: #818cf8;">P(X)</span>
</div>
<div style="display: flex; justify-content: center; gap: 28px; flex-wrap: wrap; margin-top: 10px;">
  <div style="text-align: center;">
    <div style="font-size: 18px; color: #f3e8ff; font-weight: 700;">P(y | X)</div>
    <div style="font-size: 12px; color: #a78bfa; margin-top: 4px;">Posterior<br><em>What we want</em></div>
  </div>
  <div style="text-align: center;">
    <div style="font-size: 18px; color: #c084fc; font-weight: 700;">P(X | y)</div>
    <div style="font-size: 12px; color: #a78bfa; margin-top: 4px;">Likelihood<br><em>How likely this data given class y</em></div>
  </div>
  <div style="text-align: center;">
    <div style="font-size: 18px; color: #c084fc; font-weight: 700;">P(y)</div>
    <div style="font-size: 12px; color: #a78bfa; margin-top: 4px;">Prior<br><em>How common is class y</em></div>
  </div>
  <div style="text-align: center;">
    <div style="font-size: 18px; color: #818cf8; font-weight: 700;">P(X)</div>
    <div style="font-size: 12px; color: #a78bfa; margin-top: 4px;">Evidence<br><em>Constant — we ignore it</em></div>
  </div>
</div>
</div>
**In plain English:** To decide if an email is spam, ask — *"Among all the spam emails I've seen, how often did they contain these words? And how common is spam overall?"* Multiply those two numbers together, and you have your answer.
 
Since `P(X)` is the same for all classes, we just compare the numerators and pick the class with the highest value.


---
 
<a id="naive"></a>
 
## 🤔 Why "Naive"?
 
<div style="background: linear-gradient(135deg, #1e0a3c, #2d1060); border-radius: 14px; padding: 26px 30px; margin: 24px 0; border: 1px solid #4c1d95;">
The algorithm makes one bold assumption: **all features are independent of each other**, given the class label.
 
In reality, this is almost never true. The words "free" and "win" are more likely to co-occur in a spam message — they're not independent. Yet Naive Bayes *treats them as if they are*.
 
**Why does this still work?**
 
Because we don't need perfectly calibrated probabilities — we just need the right class to rank highest. The naive assumption simplifies the math enormously, turning a complex joint probability into a chain of simple multiplications:
 
<div style="background: #0f0520; border-radius: 10px; padding: 18px 22px; margin-top: 16px; font-family: 'Georgia', serif; font-size: 16px; color: #e9d5ff; text-align: center;">
P(X | y) = P(x₁ | y) · P(x₂ | y) · P(x₃ | y) · ... · P(xₙ | y)
</div>
This makes Naive Bayes lightning-fast and trainable with very little data.
 
</div>

<a id="gaussian"></a>
 
<div style="display: flex; align-items: center; gap: 12px; margin-bottom: 8px;">
<div style="background: linear-gradient(135deg, #a855f7, #7c3aed); color: white; font-size: 12px; font-weight: 700; padding: 5px 14px; border-radius: 20px; letter-spacing: 1px;">VARIANT 1</div>
<h2 style="margin: 0; color: #f3e8ff;">GaussianNB — Iris Dataset</h2>
</div>
<div style="background: #160630; border: 1px solid #5b21b6; border-radius: 12px; padding: 20px 24px; margin-bottom: 24px;">
<strong style="color: #c084fc;">📌 Use when:</strong> <span style="color: #e9d5ff;">Features are <strong>continuous numerical values</strong> (e.g. measurements, sensor readings, physical attributes).</span>
<br><br>
 
<strong style="color: #c084fc;">💡 How it works:</strong> <span style="color: #d8b4fe;">Instead of counting occurrences, GaussianNB assumes each feature follows a <strong>normal (bell-curve) distribution</strong> within each class. It estimates the mean and variance of each feature per class during training, then uses those to compute likelihoods.</span>
</div>

The Iris dataset is a perfect fit — sepal/petal measurements are continuous, and each species (class) has its own characteristic distribution of those measurements.


In [1]:
import pandas as pd

In [2]:
iris_df = pd.read_csv("/kaggle/input/datasets/uciml/iris/Iris.csv")
iris_df

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
0,1,5.1,3.5,1.4,0.2,Iris-setosa
1,2,4.9,3.0,1.4,0.2,Iris-setosa
2,3,4.7,3.2,1.3,0.2,Iris-setosa
3,4,4.6,3.1,1.5,0.2,Iris-setosa
4,5,5.0,3.6,1.4,0.2,Iris-setosa
...,...,...,...,...,...,...
145,146,6.7,3.0,5.2,2.3,Iris-virginica
146,147,6.3,2.5,5.0,1.9,Iris-virginica
147,148,6.5,3.0,5.2,2.0,Iris-virginica
148,149,6.2,3.4,5.4,2.3,Iris-virginica


We drop the `Id` column (it carries no information) and use the four measurement columns as features:

In [3]:
X = iris_df.drop(columns=["Id", "Species"])
y = iris_df["Species"]

In [4]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    random_state=42,
    stratify=y      # keeps class proportions equal in both splits
)

`stratify=y` ensures each species is proportionally represented in both train and test — important for multi-class problems.

In [5]:
from sklearn.naive_bayes import GaussianNB

model = GaussianNB()

model.fit(X_train, y_train)

GaussianNB()

In [6]:
y_pred = model.predict(X_test)

In [7]:
from sklearn.metrics import accuracy_score, classification_report

print("Accuracy:", accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))

Accuracy: 0.9210526315789473
                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00        12
Iris-versicolor       0.86      0.92      0.89        13
 Iris-virginica       0.92      0.85      0.88        13

       accuracy                           0.92        38
      macro avg       0.92      0.92      0.92        38
   weighted avg       0.92      0.92      0.92        38



<div style="background: #0f0520; border-left: 4px solid #a855f7; border-radius: 8px; padding: 16px 20px; margin: 20px 0;">
<strong style="color: #c084fc;">📊 Reading the results:</strong>
<ul style="color: #d8b4fe; margin: 10px 0 0; line-height: 1.9;">
<li><strong style="color: #f3e8ff;">92% accuracy</strong> — excellent for a simple model with no tuning.</li>
<li><strong style="color: #f3e8ff;">Iris-setosa</strong> is perfectly classified — it's linearly separable from the others.</li>
<li>The two other species are harder to distinguish (they overlap), but Gaussian NB still handles them well.</li>
</ul>
</div>

---
 
<a id="multinomial"></a>
 
<div style="display: flex; align-items: center; gap: 12px; margin-bottom: 8px;">
<div style="background: linear-gradient(135deg, #a855f7, #7c3aed); color: white; font-size: 12px; font-weight: 700; padding: 5px 14px; border-radius: 20px; letter-spacing: 1px;">VARIANT 2</div>
<h2 style="margin: 0; color: #f3e8ff;">MultinomialNB — SMS Spam</h2>
</div>
<div style="background: #160630; border: 1px solid #5b21b6; border-radius: 12px; padding: 20px 24px; margin-bottom: 24px;">
<strong style="color: #c084fc;">📌 Use when:</strong> <span style="color: #e9d5ff;">Features represent <strong>word counts or term frequencies</strong> — any non-negative integer counts.</span>
<br><br>
 
<strong style="color: #c084fc;">💡 How it works:</strong> <span style="color: #d8b4fe;">Multinomial NB models the probability of each word appearing a given number of times. The more a word appears in spam training emails, the more it pushes the posterior toward spam. It's the workhorse of text classification.</span>
</div>
We first need to convert raw text into numerical features using **CountVectorizer**, which counts how many times each word appears in each message.
 
```python
sms_df = pd.read_csv(".../sms_spam_collection.csv")
sms_df.head()
```
 
```
  label                                            message
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
```
 
```python
# Encode labels: ham = 0, spam = 1
sms_df["label"] = sms_df["label"].map({"ham": 0, "spam": 1})
 
X = sms_df["message"]
y = sms_df["label"]
 
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)
```
 
```python
from sklearn.feature_extraction.text import CountVectorizer
 
vectorizer = CountVectorizer()
 
# fit_transform on train: learns the vocabulary AND transforms
X_train_vec = vectorizer.fit_transform(X_train)
 
# transform only on test: uses the vocabulary learned from train
X_test_vec = vectorizer.transform(X_test)
```
 
> ⚠️ **Always** `fit_transform` on training data and `transform` only on test data. Fitting on test data causes **data leakage**.
 
```python
from sklearn.naive_bayes import MultinomialNB
 
model = MultinomialNB()
model.fit(X_train_vec, y_train)
 
y_pred = model.predict(X_test_vec)
print("Accuracy:", accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))
```
 
```
Accuracy: 0.9874439461883409
 
              precision    recall  f1-score   support
 
           0       0.99      1.00      0.99       966
           1       0.99      0.92      0.95       149
 
    accuracy                           0.99      1115
```
 
<div style="background: #0f0520; border-left: 4px solid #a855f7; border-radius: 8px; padding: 16px 20px; margin: 20px 0;">
<strong style="color: #c084fc;">📊 Reading the results:</strong>
<ul style="color: #d8b4fe; margin: 10px 0 0; line-height: 1.9;">
<li><strong style="color: #f3e8ff;">98.7% accuracy</strong> — remarkable for such a simple model.</li>
<li>Note the dataset imbalance: 966 ham vs 149 spam. Despite this, MultinomialNB handles it well.</li>
<li>Spam recall is 0.92 — it misses ~8% of spam. In a real filter you'd tune the threshold to catch more spam at the cost of some false positives.</li>
</ul>
</div>
---
 
<a id="bernoulli"></a>
 
<div style="display: flex; align-items: center; gap: 12px; margin-bottom: 8px;">
<div style="background: linear-gradient(135deg, #a855f7, #7c3aed); color: white; font-size: 12px; font-weight: 700; padding: 5px 14px; border-radius: 20px; letter-spacing: 1px;">VARIANT 3</div>
<h2 style="margin: 0; color: #f3e8ff;">BernoulliNB — SMS Spam</h2>
</div>
<div style="background: #160630; border: 1px solid #5b21b6; border-radius: 12px; padding: 20px 24px; margin-bottom: 24px;">
<strong style="color: #c084fc;">📌 Use when:</strong> <span style="color: #e9d5ff;">Features are <strong>binary (0 or 1)</strong> — whether a word appears at all, not how many times.</span>
<br><br>
 
<strong style="color: #c084fc;">💡 How it works:</strong> <span style="color: #d8b4fe;">Bernoulli NB only cares about presence or absence. It also explicitly penalizes the <em>absence</em> of expected words — something MultinomialNB doesn't do. This makes it better for short texts where absence is meaningful.</span>
</div>
<div style="background: #0f0520; border-radius: 12px; padding: 20px 24px; margin: 20px 0; border: 1px solid #4c1d95;">
**MultinomialNB vs BernoulliNB — the key difference:**
 
| | Word appears 3 times | Word doesn't appear |
|---|---|---|
| **MultinomialNB** | → `3` | → `0` (ignored) |
| **BernoulliNB** | → `1` | → `0` (actively penalized) |
 
</div>
The only change from MultinomialNB is `binary=True` in CountVectorizer — it clips all counts to 1:
 
```python
vectorizer = CountVectorizer(binary=True)   # ← key difference
 
X_train_vec = vectorizer.fit_transform(X_train)
X_test_vec  = vectorizer.transform(X_test)
```
 
```python
from sklearn.naive_bayes import BernoulliNB
 
model = BernoulliNB()
model.fit(X_train_vec, y_train)
 
y_pred = model.predict(X_test_vec)
print("Accuracy:", accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))
```
 
```
Accuracy: 0.9757847533632287
 
              precision    recall  f1-score   support
 
           0       0.97      1.00      0.99       966
           1       1.00      0.82      0.90       149
```
 
<div style="background: #0f0520; border-left: 4px solid #a855f7; border-radius: 8px; padding: 16px 20px; margin: 20px 0;">
<strong style="color: #c084fc;">📊 MultinomialNB vs BernoulliNB on SMS:</strong>
<ul style="color: #d8b4fe; margin: 10px 0 0; line-height: 1.9;">
<li>MultinomialNB: <strong style="color: #f3e8ff;">98.7%</strong> accuracy, spam recall 0.92</li>
<li>BernoulliNB: <strong style="color: #f3e8ff;">97.6%</strong> accuracy, spam recall 0.82</li>
<li>MultinomialNB wins here because word frequency is informative — spam messages often repeat trigger words like "FREE" or "WINNER" multiple times.</li>
</ul>
</div>
---
 
<a id="categorical"></a>
 
<div style="display: flex; align-items: center; gap: 12px; margin-bottom: 8px;">
<div style="background: linear-gradient(135deg, #a855f7, #7c3aed); color: white; font-size: 12px; font-weight: 700; padding: 5px 14px; border-radius: 20px; letter-spacing: 1px;">VARIANT 4</div>
<h2 style="margin: 0; color: #f3e8ff;">CategoricalNB — Titanic</h2>
</div>
<div style="background: #160630; border: 1px solid #5b21b6; border-radius: 12px; padding: 20px 24px; margin-bottom: 24px;">
<strong style="color: #c084fc;">📌 Use when:</strong> <span style="color: #e9d5ff;">Features are <strong>discrete categories</strong> — labels with no inherent order (e.g. gender, passenger class, port of embarkation).</span>
<br><br>
 
<strong style="color: #c084fc;">💡 How it works:</strong> <span style="color: #d8b4fe;">CategoricalNB estimates the probability of each category value for each feature, given each class. It handles <code>male/female</code> or <code>1st/2nd/3rd</code> class natively — no need to treat them as numbers.</span>
</div>
The Titanic dataset is a natural fit — `Sex`, `Pclass`, and `Embarked` are all categorical. The preprocessing steps are:
 
1. Select only categorical features
2. Drop rows with missing values (or impute)
3. Encode string categories as integers using `OrdinalEncoder`
```python
titanic_df = pd.read_csv(".../Titanic_Survival_Dataset.csv")
 
# Select categorical features
features = ["Pclass", "Sex", "Embarked"]
titanic_clean = titanic_df[features + ["Survived"]].dropna()
 
X = titanic_clean[features]
y = titanic_clean["Survived"]
```
 
```python
from sklearn.preprocessing import OrdinalEncoder
from sklearn.naive_bayes import CategoricalNB
 
encoder = OrdinalEncoder()
X_encoded = encoder.fit_transform(X)
 
X_train, X_test, y_train, y_test = train_test_split(
    X_encoded, y, test_size=0.2, random_state=42, stratify=y
)
 
model = CategoricalNB()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
 
print("Accuracy:", accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))
```
 
> 💡 **Note:** `OrdinalEncoder` converts `male/female` → `0/1`, but CategoricalNB interprets these as distinct *categories*, not as a numerical ordering — that's what makes it the right tool here.
 
---
 
<a id="complement"></a>
 
<div style="display: flex; align-items: center; gap: 12px; margin-bottom: 8px;">
<div style="background: linear-gradient(135deg, #a855f7, #7c3aed); color: white; font-size: 12px; font-weight: 700; padding: 5px 14px; border-radius: 20px; letter-spacing: 1px;">VARIANT 5</div>
<h2 style="margin: 0; color: #f3e8ff;">ComplementNB — SMS Spam</h2>
</div>
<div style="background: #160630; border: 1px solid #5b21b6; border-radius: 12px; padding: 20px 24px; margin-bottom: 24px;">
<strong style="color: #c084fc;">📌 Use when:</strong> <span style="color: #e9d5ff;">Your dataset is <strong>imbalanced</strong> and features are counts (text). It's a smarter sibling of MultinomialNB for skewed class distributions.</span>
<br><br>
 
<strong style="color: #c084fc;">💡 How it works:</strong> <span style="color: #d8b4fe;">Instead of modeling <em>"how likely is this word given spam?"</em>, ComplementNB models <em>"how likely is this word given everything that is NOT spam?"</em> It estimates the complement of each class and uses that to classify. This correction makes it more robust when one class dominates the training data.</span>
</div>
The implementation mirrors MultinomialNB exactly — the only change is the model class:
 
```python
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import ComplementNB
 
vectorizer = CountVectorizer()
X_train_vec = vectorizer.fit_transform(X_train)
X_test_vec  = vectorizer.transform(X_test)
 
model = ComplementNB()      # ← the only difference
model.fit(X_train_vec, y_train)
 
y_pred = model.predict(X_test_vec)
print("Accuracy:", accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))
```
 
```
Accuracy: 0.9829596412556054
 
              precision    recall  f1-score   support
 
           0       0.99      0.99      0.99       966
           1       0.93      0.95      0.94       149
```
 
<div style="background: #0f0520; border-left: 4px solid #a855f7; border-radius: 8px; padding: 16px 20px; margin: 20px 0;">
<strong style="color: #c084fc;">📊 ComplementNB vs MultinomialNB:</strong>
<ul style="color: #d8b4fe; margin: 10px 0 0; line-height: 1.9;">
<li>ComplementNB spam recall: <strong style="color: #f3e8ff;">0.95</strong> vs MultinomialNB's <strong style="color: #f3e8ff;">0.92</strong> — it catches 3% more spam.</li>
<li>On more severely imbalanced datasets, ComplementNB often outperforms MultinomialNB more dramatically.</li>
<li>When in doubt with text classification and class imbalance, ComplementNB is the safer default.</li>
</ul>
</div>
---
 
<a id="cheatsheet"></a>
 
## 📊 Comparison & Cheat Sheet
 
<div style="background: linear-gradient(135deg, #1a0533, #2d1060); border-radius: 16px; padding: 30px; margin: 30px 0;">
<h3 style="color: #e9d5ff; margin: 0 0 24px; text-align: center;">Which Naive Bayes should I use?</h3>
 
| Variant | Input Features | Best For | Example |
|---------|---------------|----------|---------|
| **GaussianNB** | Continuous numbers | Physical measurements, sensor data | 🌸 Iris species |
| **MultinomialNB** | Word counts (integers) | Text classification, NLP | 📱 Spam detection |
| **BernoulliNB** | Binary 0/1 | Short texts, document presence | 📱 Spam detection |
| **CategoricalNB** | Categorical labels | Tabular data with categories | 🚢 Titanic survival |
| **ComplementNB** | Word counts (imbalanced) | Imbalanced text datasets | 📱 Spam (minority class) |
 
</div>
<div style="background: #0f0520; border-radius: 16px; padding: 30px; margin: 24px 0; border: 1px solid #4c1d95; font-family: monospace; font-size: 14px; color: #d8b4fe; line-height: 1.7;">
<pre style="margin: 0; color: #e9d5ff; overflow-x: auto;">
                    Naive Bayes
                         │
        ┌────────────────┼─────────────────┐
        │                │                 │
   Numerical          Text             Categorical
        │                │                 │
  GaussianNB       ┌─────┴──────┐     CategoricalNB
                   │            │
          ┌────────┴───┐    BernoulliNB
          │            │    (binary 0/1)
  MultinomialNB  ComplementNB
   (counts)      (imbalanced)</pre>
 
</div>
---
 
<div style="background: linear-gradient(135deg, #1a0533, #3b0764, #4a1a8a); border-radius: 16px; padding: 36px 32px; margin-top: 40px; border: 1px solid #6b21a8;">
<h3 style="color: #e9d5ff; margin: 0 0 20px; font-size: 22px;">🧠 Key Takeaways</h3>
 
<div style="display: grid; gap: 14px;">
<div style="background: rgba(168,85,247,0.15); border-radius: 10px; padding: 14px 18px; border-left: 3px solid #a855f7;">
<span style="color: #c084fc; font-weight: 700;">The naive assumption is a feature, not a bug.</span>
<span style="color: #d8b4fe;"> Ignoring feature interactions makes Naive Bayes blazing fast and surprisingly accurate — good enough for many real-world problems.</span>
</div>
<div style="background: rgba(168,85,247,0.15); border-radius: 10px; padding: 14px 18px; border-left: 3px solid #a855f7;">
<span style="color: #c084fc; font-weight: 700;">Match the variant to your data type.</span>
<span style="color: #d8b4fe;"> Gaussian for numbers, Multinomial/Bernoulli/Complement for text, Categorical for label features — using the wrong variant is the most common mistake.</span>
</div>
<div style="background: rgba(168,85,247,0.15); border-radius: 10px; padding: 14px 18px; border-left: 3px solid #a855f7;">
<span style="color: #c084fc; font-weight: 700;">Always fit on train, transform on test.</span>
<span style="color: #d8b4fe;"> Whether it's a vectorizer or an encoder, fitting on the test set leaks information and inflates your metrics.</span>
</div>
<div style="background: rgba(168,85,247,0.15); border-radius: 10px; padding: 14px 18px; border-left: 3px solid #a855f7;">
<span style="color: #c084fc; font-weight: 700;">ComplementNB is your imbalanced-data insurance.</span>
<span style="color: #d8b4fe;"> When one class dominates, it corrects for that bias and typically recovers better recall on the minority class.</span>
</div>
</div>
<div style="margin-top: 28px; padding-top: 20px; border-top: 1px solid rgba(168,85,247,0.3); text-align: center;">
<span style="color: #a78bfa; font-size: 14px;">Episode 43 complete · Next up: Support Vector Machines →</span>
</div>
</div>

In [8]:
sms_df = pd.read_csv(
    "/kaggle/input/datasets/hamnawaseem112222222/sms-spam-collection-5572-labeled-sms-messages/sms_spam_collection (1).csv"
)

sms_df.head()

,label,message
0,ham,"Go until jurong point, crazy.. Available only ..."
1,ham,Ok lar... Joking wif u oni...
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...
3,ham,U dun say so early hor... U c already then say...
4,ham,"Nah I don't think he goes to usf, he lives aro..."


In [9]:
sms_df["label"] = sms_df["label"].map({
    "ham": 0,
    "spam": 1
})

sms_df.head()

,label,message
0,0,"Go until jurong point, crazy.. Available only ..."
1,0,Ok lar... Joking wif u oni...
2,1,Free entry in 2 a wkly comp to win FA Cup fina...
3,0,U dun say so early hor... U c already then say...
4,0,"Nah I don't think he goes to usf, he lives aro..."


In [10]:
X = sms_df["message"]
y = sms_df["label"]

In [11]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [12]:
from sklearn.feature_extraction.text import CountVectorizer

vectorizer = CountVectorizer()

X_train_vec = vectorizer.fit_transform(X_train)
X_test_vec = vectorizer.transform(X_test)

In [13]:
from sklearn.naive_bayes import MultinomialNB

model = MultinomialNB()

model.fit(X_train_vec, y_train)

MultinomialNB()

In [14]:
y_pred = model.predict(X_test_vec)

In [15]:
from sklearn.metrics import accuracy_score, classification_report

print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nClassification Report:\n")
print(classification_report(y_test, y_pred))

Accuracy: 0.9874439461883409

Classification Report:

              precision    recall  f1-score   support

           0       0.99      1.00      0.99       966
           1       0.99      0.92      0.95       149

    accuracy                           0.99      1115
   macro avg       0.99      0.96      0.97      1115
weighted avg       0.99      0.99      0.99      1115



# 3. BernoulliNB — SMS Spam

MultinomialNB:

`word appears 3 times → 3`

BernoulliNB:

`word appears → 1`
`word doesn't appear → 0`

In [16]:
vectorizer = CountVectorizer(binary=True)

X_train_vec = vectorizer.fit_transform(X_train)
X_test_vec = vectorizer.transform(X_test)

In [17]:
from sklearn.naive_bayes import BernoulliNB

model = BernoulliNB()

model.fit(X_train_vec, y_train)

BernoulliNB()

In [18]:
y_pred = model.predict(X_test_vec)

print("Accuracy:", accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))

Accuracy: 0.9757847533632287
              precision    recall  f1-score   support

           0       0.97      1.00      0.99       966
           1       1.00      0.82      0.90       149

    accuracy                           0.98      1115
   macro avg       0.99      0.91      0.94      1115
weighted avg       0.98      0.98      0.97      1115



# 4. CategoricalNB — Titanic

**Use when:** your features are categorical variables.

In [19]:
titanic_df = pd.read_csv("/kaggle/input/datasets/ahmedfakhar123/titanic-survival-dataset/Titanic_Survival_Dataset.csv")
titanic_df

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


# 5. ComplementNB — SMS Spam

`ComplementNB` is designed particularly with imbalanced datasets in mind and is commonly useful for text classification.

In [20]:
from sklearn.feature_extraction.text import CountVectorizer

vectorizer = CountVectorizer()

X_train_vec = vectorizer.fit_transform(X_train)
X_test_vec = vectorizer.transform(X_test)

In [21]:
from sklearn.naive_bayes import ComplementNB

model = ComplementNB()

model.fit(X_train_vec, y_train)

ComplementNB()

In [22]:
y_pred = model.predict(X_test_vec)

In [23]:
print("Accuracy:", accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))

Accuracy: 0.9829596412556054
              precision    recall  f1-score   support

           0       0.99      0.99      0.99       966
           1       0.93      0.95      0.94       149

    accuracy                           0.98      1115
   macro avg       0.96      0.97      0.96      1115
weighted avg       0.98      0.98      0.98      1115



| Naive Bayes       | Input features       | Example    |
| ----------------- | -------------------- | ---------- |
| **GaussianNB**    | Continuous numerical | 🌸 Iris    |
| **MultinomialNB** | Counts/frequencies   | 📱 SMS     |
| **BernoulliNB**   | Binary 0/1           | 📱 SMS     |
| **CategoricalNB** | Categorical          | 🚢 Titanic |
| **ComplementNB**  | Counts/frequencies   | 📱 SMS     |


```text
                    Naive Bayes
                         │
        ┌────────────────┼─────────────────┐
        │                │                 │
   Numerical          Text             Categorical
        │                │                 │
  GaussianNB       ┌─────┴─────┐      CategoricalNB
                   │           │
             MultinomialNB  BernoulliNB
                   │           │
                Counts      0 / 1
```